src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocksHelper.cs

```diff
@@ -0,0 +1,371 @@
+                buffer[8 + usernameLength] = 0;
+                int totalLength = 9 + usernameLength;
+
+                if (ipv4Address is null)
+                {
+                    // https://www.openssh.com/txt/socks4a.protocol
+                    byte hostLength = EncodeString(host, buffer.AsSpan(totalLength), nameof(host));
+                    buffer[totalLength + hostLength] = 0;
+                    totalLength += hostLength + 1;
+                }
+
+                await WriteAsync(stream, buffer.AsMemory(0, totalLength), async).ConfigureAwait(false);
+
+                // +----+----+----+----+----+----+----+----+
+                // | VN | CD | DSTPORT |      DSTIP        |
+                // +----+----+----+----+----+----+----+----+
+                //    1    1      2              4
+                await ReadToFillAsync(stream, buffer.AsMemory(0, 8), async).ConfigureAwait(false);
+
+                switch (buffer[1])
+                {
+                    case Socks4_Success:
+                        // Nothing to do
+                        break;
+                    case Socks4_AuthFailed:
+                        throw new SocksException(SR.net_socks_auth_failed);
+                    default:
+                        throw new SocksException(SR.net_socks_connection_failed);
+                }
+                // response address not used
+            }
+            finally
+            {
+                ArrayPool<byte>.Shared.Return(buffer);
+            }
+        }
+
+        private static byte EncodeString(ReadOnlySpan<char> chars, Span<byte> buffer, string parameterName)
+        {
+            try
+            {
+                return checked((byte)Encoding.UTF8.GetBytes(chars, buffer));
+            }
+            catch
+            {
+                Debug.Assert(Encoding.UTF8.GetByteCount(chars) > 255);
+                throw new SocksException(SR.Format(SR.net_socks_string_too_long, parameterName));
+            }
+        }
+
+        private static void VerifyProtocolVersion(byte expected, byte version)
+        {
+            if (expected != version)
+            {
+                throw new SocksException(SR.Format(SR.net_socks_unexpected_version, expected, version));
+            }
+        }
+
+        private static ValueTask WriteAsync(Stream stream, Memory<byte> buffer, bool async)
+        {
+            if (async)
+            {
+                return stream.WriteAsync(buffer);
+            }
+            else
+            {
+                stream.Write(buffer.Span);
+                return default;
+            }
+        }
+
+        private static async ValueTask ReadToFillAsync(Stream stream, Memory<byte> buffer, bool async)
+        {
+            while (buffer.Length != 0)
+            {
+                int bytesRead = async
+                    ? await stream.ReadAsync(buffer).ConfigureAwait(false)
```

Comment:
Should this have cancellation support?

Reply:
This is using the same approach as `HttpConnection`, where we register for cancellation once and have the callback dispose the stream. All other calls to stream Read/Write don't pass the CT and instead rely on the Dispose to wake them up.

https://github.com/dotnet/runtime/blob/95400e21568c1c1c583235e5e49dc00d911222ed/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocksHelper.cs#L34